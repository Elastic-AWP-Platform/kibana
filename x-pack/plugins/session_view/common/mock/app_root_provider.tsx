/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import React, { memo, ReactNode, useMemo } from 'react';
import { I18nProvider } from '@kbn/i18n/react';
import { Router } from 'react-router-dom';
import { History } from 'history';
import useObservable from 'react-use/lib/useObservable';
import { EuiThemeProvider } from '../../../../../src/plugins/kibana_react/common';
// eslint-disable-next-line @kbn/eslint/no-restricted-paths
import { KibanaContextProvider } from '../../../../../src/plugins/kibana_react/public';
import { CoreStart } from '../../../../../src/core/public';

/**
 * Provides the context for rendering the endpoint app
 */
export const AppRootProvider = memo<{
  history: History;
  coreStart: CoreStart;
  children: ReactNode | ReactNode[];
}>(({ history, coreStart: { http, notifications, uiSettings, application }, children }) => {
  const isDarkMode = useObservable<boolean>(uiSettings.get$('theme:darkMode'));
  const services = useMemo(
    () => ({ http, notifications, application }),
    [application, http, notifications]
  );
  return (
    <I18nProvider>
      <KibanaContextProvider services={services}>
        <EuiThemeProvider darkMode={isDarkMode}>
          <Router history={history}>{children}</Router>
        </EuiThemeProvider>
      </KibanaContextProvider>
    </I18nProvider>
  );
});

AppRootProvider.displayName = 'AppRootProvider';
