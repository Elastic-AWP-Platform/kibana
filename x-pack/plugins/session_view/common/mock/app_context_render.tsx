/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import React from 'react';
import { createMemoryHistory, MemoryHistory } from 'history';
import { render as reactRender, RenderOptions, RenderResult } from '@testing-library/react';
import { QueryClient, QueryClientProvider, setLogger } from 'react-query';
import { coreMock } from '../../../../../src/core/public/mocks';
// eslint-disable-next-line @kbn/eslint/no-restricted-paths
import { KibanaContextProvider } from '../../../../../src/plugins/kibana_react/public';
import { AppRootProvider } from './app_root_provider';

type UiRender = (ui: React.ReactElement, options?: RenderOptions) => RenderResult;

// hide react-query output in console
setLogger({
  error: () => {},
  // eslint-disable-next-line no-console
  log: console.log,
  // eslint-disable-next-line no-console
  warn: console.warn,
});

/**
 * Mocked app root context renderer
 */
export interface AppContextTestRender {
  history: ReturnType<typeof createMemoryHistory>;
  coreStart: ReturnType<typeof coreMock.createStart>;
  /**
   * A wrapper around `AppRootContext` component. Uses the mocked modules as input to the
   * `AppRootContext`
   */
  AppWrapper: React.FC<any>;
  /**
   * Renders the given UI within the created `AppWrapper` providing the given UI a mocked
   * endpoint runtime context environment
   */
  render: UiRender;
}

/**
 * Creates a mocked endpoint app context custom renderer that can be used to render
 * component that depend upon the application's surrounding context providers.
 * Factory also returns the content that was used to create the custom renderer, allowing
 * for further customization.
 */
export const createAppRootMockRenderer = (): AppContextTestRender => {
  const history = createMemoryHistory<never>();
  const coreStart = createCoreStartMock(history);

  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        // turns retries off
        retry: false,
        // prevent jest did not exit errors
        cacheTime: Infinity,
      },
    },
  });

  const AppWrapper: React.FC<{ children: React.ReactElement }> = ({ children }) => (
    <KibanaContextProvider>
      <AppRootProvider history={history} coreStart={coreStart}>
        <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
      </AppRootProvider>
    </KibanaContextProvider>
  );

  const render: UiRender = (ui, options) => {
    return reactRender(ui, {
      wrapper: AppWrapper as React.ComponentType,
      ...options,
    });
  };

  return {
    history,
    coreStart,
    AppWrapper,
    render,
  };
};

const createCoreStartMock = (
  history: MemoryHistory<never>
): ReturnType<typeof coreMock.createStart> => {
  const coreStart = coreMock.createStart({ basePath: '/mock' });

  // Mock the certain APP Ids returned by `application.getUrlForApp()`
  coreStart.application.getUrlForApp.mockImplementation((appId) => {
    switch (appId) {
      case 'sessionView':
        return '/app/sessionView';
      default:
        return `${appId} not mocked!`;
    }
  });

  coreStart.application.navigateToUrl.mockImplementation((url) => {
    history.push(url.replace('/app/sessionView', ''));
    return Promise.resolve();
  });

  return coreStart;
};
