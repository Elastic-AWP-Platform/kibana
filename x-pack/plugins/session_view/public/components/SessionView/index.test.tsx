/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { waitFor, waitForElementToBeRemoved } from '@testing-library/react';
import React from 'react';
import { sessionViewProcessEventsMock } from '../../../common/mocks/responses/session_view_process_events.mock';
import { AppContextTestRender, createAppRootMockRenderer } from '../../test';
import { SessionView } from './index';
import userEvent from '@testing-library/user-event';

describe('SessionView component', () => {
  let render: (props?: any) => ReturnType<AppContextTestRender['render']>;
  let renderResult: ReturnType<typeof render>;
  let mockedContext: AppContextTestRender;
  let mockedApi: AppContextTestRender['coreStart']['http']['get'];

  const waitForApiCall = () => waitFor(() => expect(mockedApi).toHaveBeenCalled());

  beforeEach(() => {
    mockedContext = createAppRootMockRenderer();
    mockedApi = mockedContext.coreStart.http.get;
    render = (props = {}) =>
      (renderResult = mockedContext.render(
        <SessionView sessionEntityId="ae06c110-ad2d-5830-b47c-08ad62f1734c" {...props} />
      ));
  });

  describe('When SessionView is mounted', () => {
    describe('And no data exists', () => {
      beforeEach(async () => {
        mockedApi.mockResolvedValue({
          events: [],
        });
      });

      it('should show the Empty message', async () => {
        render();
        await waitForApiCall();
        expect(renderResult.getByTestId('sessionViewProcessEventsEmpty')).toBeTruthy();
      });

      it('should not display the search bar', async () => {
        render();
        await waitForApiCall();
        expect(renderResult.queryByTestId('sessionViewProcessEventsSearch')).toBeFalsy();
      });
    });

    describe('And data exists', () => {
      beforeEach(async () => {
        mockedApi.mockResolvedValue(sessionViewProcessEventsMock);
      });

      it('should show loading indicator while retrieving data and hide it when it gets it', async () => {
        let releaseApiResponse: (value?: unknown) => void;

        // make the request wait
        mockedApi.mockReturnValue(new Promise((resolve) => (releaseApiResponse = resolve)));
        render();
        await waitForApiCall();

        // see if loader is present
        expect(renderResult.getByTestId('sectionLoading')).toBeTruthy();

        // release the request
        releaseApiResponse!(mockedApi);

        //  check the loader is gone
        await waitForElementToBeRemoved(renderResult.getByTestId('sectionLoading'));
      });

      it('should display the search bar', async () => {
        render();
        await waitForApiCall();
        expect(renderResult.getByTestId('sessionViewProcessEventsSearch')).toBeTruthy();
      });

      it('should show items on the list, and auto selects session leader', async () => {
        render();
        await waitForApiCall();

        expect(renderResult.getAllByTestId('processTreeNode')[0]).toMatchSnapshot();
      });

      it('should toggle detail panel visibilty when detail button clicked', async () => {
        render();
        await waitForApiCall();

        userEvent.click(renderResult.getByTestId('sessionViewDetailPanelToggle'));
        expect(renderResult.getByTestId('sessionViewDetailPanel')).toBeTruthy();
      });

      it('should jump To Event', async () => {
        const jumpToEvent = {
          '@timestamp': new Date('2021-11-23T13:41:46.768Z'),
          process: {
            entity_id: '85752b94-1c86-5540-9a61-743429d5a206',
          },
        };

        const sessionView = render({ jumpToEvent });
        await waitForApiCall();
        expect(
          sessionView.getByTestId(`processTreeNodeRow-${jumpToEvent.process.entity_id}`)
        ).toBeInTheDocument();
      });
    });
  });
});
