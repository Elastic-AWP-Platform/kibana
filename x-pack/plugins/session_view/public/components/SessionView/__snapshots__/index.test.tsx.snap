// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionView component When SessionView is mounted And data exists should show items on the list, and auto selects session leader 1`] = `
.emotion-0 {
  line-height: 18px;
  height: 20px;
  font-size: 11px;
  font-family: 'Roboto Mono',Menlo,Courier,monospace;
  border-radius: 6px;
  color: #343741;
  margin-left: 8px;
  min-width: 0;
  background: rgba(170, 101, 86, 0.04);
  border: 1px solid rgba(170, 101, 86, 0.48);
}

.emotion-1 {
  margin-left: 8px;
}

<div
  css="[object Object]"
  data-id="ae06c110-ad2d-5830-b47c-08ad62f1734c"
  data-test-subj="processTreeNode"
>
  <div
    css="[object Object]"
    data-test-subj="processTreeNodeRow-ae06c110-ad2d-5830-b47c-08ad62f1734c"
  >
    <span
      data-euiicon-type="consoleApp"
    />
     
    <b
      css="[object Object]"
    >
      cat
    </b>
     
    started by
     
    <span
      data-euiicon-type="user"
    />
     
    <b
      css="[object Object]"
    >
      vagrant
    </b>
    <span
      class="euiToolTipAnchor"
    >
      <button
        class="euiButton euiButton--primary emotion-0"
        data-test-subj="processTreeNodeChildProcessesButton"
        type="button"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <span
            class="euiButton__text"
          >
            +2 more
            <span
              class="emotion-1"
              data-euiicon-type="arrowDown"
            />
          </span>
        </span>
      </button>
    </span>
  </div>
</div>
`;
